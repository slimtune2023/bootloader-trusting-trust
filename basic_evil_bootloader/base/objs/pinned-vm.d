objs/pinned-vm.o: pinned-vm.c \
 /Users/jameschen/Documents/cs140e-25win/libpi/include/rpi.h \
 /Users/jameschen/Documents/cs140e-25win/libpi/libc/demand.h \
 /Users/jameschen/Documents/cs140e-25win/libpi/include/gpio.h \
 /Users/jameschen/Documents/cs140e-25win/libpi/include/rpi-constants.h \
 /Users/jameschen/Documents/cs140e-25win/libpi/include/your-prototypes.h \
 /Users/jameschen/Documents/cs140e-25win/libpi/libc/bit-support.h \
 /Users/jameschen/Documents/cs140e-25win/libpi/src/vector-base.h \
 /Users/jameschen/Documents/cs140e-25win/libpi/include/asm-helpers.h \
 pinned-vm.h mmu.h armv6-cp15.h mem-attr.h armv6-coprocessor-asm.h \
 procmap.h memmap-default.h \
 /Users/jameschen/Documents/cs140e-25win/libpi/include/memmap.h

/Users/jameschen/Documents/cs140e-25win/libpi/include/rpi.h:

/Users/jameschen/Documents/cs140e-25win/libpi/libc/demand.h:

/Users/jameschen/Documents/cs140e-25win/libpi/include/gpio.h:

/Users/jameschen/Documents/cs140e-25win/libpi/include/rpi-constants.h:

/Users/jameschen/Documents/cs140e-25win/libpi/include/your-prototypes.h:

/Users/jameschen/Documents/cs140e-25win/libpi/libc/bit-support.h:

/Users/jameschen/Documents/cs140e-25win/libpi/src/vector-base.h:

/Users/jameschen/Documents/cs140e-25win/libpi/include/asm-helpers.h:

pinned-vm.h:

mmu.h:

armv6-cp15.h:

mem-attr.h:

armv6-coprocessor-asm.h:

procmap.h:

memmap-default.h:

/Users/jameschen/Documents/cs140e-25win/libpi/include/memmap.h:
