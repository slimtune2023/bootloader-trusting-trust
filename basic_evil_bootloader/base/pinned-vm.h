#ifndef __PINNED_VM_H__
#define __PINNED_VM_H__
// simple interface for pinned virtual memory: most of it is 
// enum and data structures.  you'll build three routines:

#include "mmu.h"
#include "mem-attr.h"
#include "asm-helpers.h"
#include "armv6-coprocessor-asm.h"

// you can flip these back and forth if you want debug output.
#if 0
    // change <output> to <debug> if you want file/line/func
#   define pin_debug(args...) output("PIN_VM:" args)
#else
#   define pin_debug(args...) do { } while(0)
#endif

#define coproc_mk_set(fn_name, coproc, opcode_1, Crn, Crm, opcode_2)       \
    static inline void c ## coproc ## _ ## fn_name ## _set(uint32_t v) {                    \
        asm volatile ("mcr " MK_STR(coproc) ", "                        \
                             MK_STR(opcode_1) ", "                      \
                             "%0, "                                     \
                            MK_STR(Crn) ", "                            \
                            MK_STR(Crm) ", "                            \
                            MK_STR(opcode_2) :: "r" (v));               \
        prefetch_flush();                                               \
    }

#define coproc_mk_get(fn_name, coproc, opcode_1, Crn, Crm, opcode_2)       \
    static inline uint32_t c ## coproc ## _ ## fn_name ## _get(void) {                      \
        uint32_t ret=0;                                                   \
        asm volatile ("mrc " MK_STR(coproc) ", "                        \
                             MK_STR(opcode_1) ", "                      \
                             "%0, "                                     \
                            MK_STR(Crn) ", "                            \
                            MK_STR(Crm) ", "                            \
                            MK_STR(opcode_2) : "=r" (ret));             \
        return ret;                                                     \
    }


// make both get and set methods.
#define coproc_mk(fn, coproc, opcode_1, Crn, Crm, opcode_2)     \
    coproc_mk_set(fn, coproc, opcode_1, Crn, Crm, opcode_2)        \
    coproc_mk_get(fn, coproc, opcode_1, Crn, Crm, opcode_2) 

coproc_mk(tlb_idx, p15, 5, c15, c4, 2)
coproc_mk(tlb_va, p15, 5, c15, c5, 2)
coproc_mk(tlb_pa, p15, 5, c15, c6, 2)
coproc_mk(tlb_attr, p15, 5, c15, c7, 2)
coproc_mk_set(va_trans, p15, 0, c7, c8, 0) // op code 2?
coproc_mk_get(pa_trans, p15, 0, c7, c4, 0)

// attributes: these get inserted into the TLB.
typedef struct {
    // for today we only handle 1MB sections.
    uint32_t G,         // is this a global entry?
             asid,      // don't think this should actually be in this.
             dom,       // domain id
             pagesize;  // can be 1MB or 16MB

    // permissions for needed to access page.
    //
    // see mem_perm_t above: is the bit merge of 
    // APX and AP so can or into AP position.
    mem_perm_t  AP_perm;

    // caching policy for this memory.
    // 
    // see mem_cache_t enum above.
    // see table on 6-16 b4-8/9
    // for today everything is uncacheable.
    mem_attr_t mem_attr;
} pin_t;

// pinned encodings for different page sizes.
enum {
    PAGE_4K     = 0b01,
    PAGE_64K    = 0b10,
    PAGE_1MB    = 0b11,
    PAGE_16MB   = 0b00
};

static inline pin_t pin_16mb(pin_t p) {
    p.pagesize = PAGE_16MB;
    return p;
}

// enable MMU -- must have set the context
// first.
void pin_mmu_enable(void);

// disable MMU
void pin_mmu_disable(void);

// mmu can be off or on: setup address space context.
// must do before enabling MMU or when switching
// between address spaces.
void pin_set_context(uint32_t asid);

// our main routine: 
//  insert map <va> ==> <pa> with attributes <attr> 
//  at position <idx> in the TLB.
//
// errors:
//  - idx >= 8.
//  - va already mapped.
void pin_mmu_sec(unsigned idx,
                uint32_t va,
                uint32_t pa,
                pin_t attr);

// do a manual translation in tlb and see if exists (1)
// returns the result in <result>
//
// low 1..6 bits of result should have the reason.
int tlb_contains_va(uint32_t *result, uint32_t va);

// wrapper that check that <va> is pinned in the tlb
int pin_exists(uint32_t va, int verbose_p);

// print <msg> then all valid pinned entries.
// note: if you print everything you see a lot of
// garbage in the non-initialized entires --- i'm 
// not sure if we are guaranteed that these have their
// valid bit set to 0?   
void lockdown_print_entries(const char* msg);

// set pin index <idx> to all 0s.
void pin_clear(unsigned idx);

void pin_mmu_init(uint32_t domain_reg);

// void staff_pin_mmu_init(uint32_t domain_reg);

static inline pin_t pin_mk_global(uint32_t dom, mem_perm_t perm, mem_attr_t attr) {
    pin_t p;
    p.G = 1;
    p.asid = 0;
    p.dom = dom;
    p.pagesize = 0b11;
    p.AP_perm = perm;
    p.mem_attr = attr;
    return p;
}

static inline pin_t pin_mk_user(uint32_t dom, uint32_t asid, mem_perm_t perm, mem_attr_t attr) {
    pin_t p;
    p.G = 0;
    p.asid = asid;
    p.dom = dom;
    p.pagesize = 0b11;
    p.AP_perm = perm;
    p.mem_attr = attr;
    return p;
}

static inline pin_t pin_mk_device(uint32_t dom) {
    pin_t p;
    p.G = 1;
    p.asid = 0;  // unset
    p.dom = dom;
    p.pagesize = 0b11;
    p.AP_perm = perm_rw_priv;
    p.mem_attr = MEM_device;
    return p;
}

#endif