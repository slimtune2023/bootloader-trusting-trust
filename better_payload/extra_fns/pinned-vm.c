#include "rpi.h"
#include "libc/bit-support.h"

// has useful enums and helpers.
#include "vector-base.h"
#include "pinned-vm.h"
#include "mmu.h"
#include "procmap.h"

// generate the _get and _set methods.
// (see asm-helpers.h for the cp_asm macro 
// definition)
// arm1176.pdf: 3-149

// static void *null_pt = 0;

void pin_mmu_enable(void) {
    mmu_enable();
}

void pin_mmu_disable(void) {
    mmu_disable();
}

void pin_set_context(uint32_t asid) {
    void *null_pt = kmalloc_aligned(4096*4, 1<<14);
    mmu_set_ctx(128, asid, null_pt);
}

// our main routine: 
//  insert map <va> ==> <pa> with attributes <attr> 
//  at position <idx> in the TLB.
//
// errors:
//  - idx >= 8.
//  - va already mapped.
void pin_mmu_sec(unsigned idx, uint32_t va, uint32_t pa, pin_t attr) {
    assert(idx < 8);
    uint32_t *temp = 0;
    // assert(!tlb_contains_va(temp, va));

    cp15_tlb_idx_set(idx);


    // printk("asid: %x\n", attr.asid);
    unsigned va_set = attr.asid;
    va_set |= attr.G << 9;
    va_set |= va;
    cp15_tlb_va_set(va_set);

    unsigned attr_set = attr.mem_attr << 1;
    attr_set |= attr.dom << 7;
    cp15_tlb_attr_set(attr_set);

    unsigned pa_set = 1;
    pa_set |= attr.AP_perm << 1;
    pa_set |= attr.pagesize << 6; // size; change if not 1mb
    pa_set |= 0b00 << 8;
    pa_set |= pa;
    cp15_tlb_pa_set(pa_set);

    asm volatile ("mcr p15, 0, r0, c7, c10, 4");  // DSB
    asm volatile ("mcr p15, 0, r0, c7, c5, 4");   // prefetch_flush

    // assert(tlb_contains_va(temp, va));
    // branch thingy?
}

// do a manual translation in tlb and see if exists (1)
// returns the result in <result>
//
// low 1..6 bits of result should have the reason.
int tlb_contains_va(uint32_t *result, uint32_t va) {
    cp15_va_trans_set(va);
    *result = cp15_pa_trans_get();
    uint32_t ret = 1-(*result & 1);
    *result &= ~0x3ff;
    *result |= (va & 0x3ff);
    return ret;
}

int pin_exists(uint32_t va, int verbose_p) {
    uint32_t res;
    return tlb_contains_va(&res, va);
}

uint32_t* get_regs_idx(unsigned idx) {
    cp15_tlb_idx_set(idx);

    unsigned va = cp15_tlb_va_get();
    unsigned pa = cp15_tlb_pa_get();
    unsigned attr = cp15_tlb_attr_get();

    asm volatile ("mcr p15, 0, r0, c7, c10, 4");  // DSB
    asm volatile ("mcr p15, 0, r0, c7, c5, 4");   // prefetch_flush

    uint32_t* ret = 0;
    *ret = va;
    *(ret+1) = pa;
    *(ret+2) = attr;
    return ret;
}

// void lockdown_print_entries(const char* msg) {
//     printk("%s\n", msg);
//     unsigned* ret;
//     uint32_t* res = 0;
//     for (int i = 0; i < 8; i++) {
//         ret = get_regs_idx(i);
//         printk("TLB entry %d:\n", i);
//         printk("va: %x\t", ret[0]);
//         printk("pa: %x\t", ret[1]);
//         printk("attr: %x\t", ret[2]);
//         tlb_contains_va(res, ret[0]);
//         printk("contains: %x\n", *res);
//     }
// }

void lockdown_print_entry(unsigned idx) {
    trace("   idx=%d\n", idx);
    // lockdown_index_set(idx);
    // uint32_t va_ent = lockdown_va_get();
    // uint32_t pa_ent = lockdown_pa_get();

    uint32_t *ret = get_regs_idx(idx);
    uint32_t va_ent = ret[0];
    uint32_t pa_ent = ret[1];
    uint32_t attr = ret[2];

    uint32_t v = pa_ent & 0x1; 

    if(!v) {
        trace("     [invalid entry %d]\n", idx);
        return;
    }

    // 3-149
    uint32_t va = (va_ent >> 12);
    uint32_t G = (va_ent >> 9) & 1;
    uint32_t asid = va_ent & 0xff;
    trace("     va_ent=%x: va=%x|G=%d|ASID=%d\n",
        va_ent, va, G, asid);

    // 3-150
    uint32_t pa = (pa_ent >> 12);
    uint32_t nsa = (pa_ent >> 9) & 1;
    uint32_t nstid = (pa_ent >> 8) & 1;
    uint32_t size = (pa_ent >> 6) & 0x3; 
    uint32_t apx = (pa_ent >> 1) & 0x3;
    trace("     pa_ent=%x: pa=%x|nsa=%d|nstid=%d|size=%b|apx=%b|v=%d\n",
                pa_ent, pa, nsa,nstid,size, apx,v);

    // 3-151
    uint32_t dom = (attr >> 7) & 0xf;
    uint32_t xn = (attr >> 6) & 1;
    uint32_t tex = (attr >> 3) & 0x7; 
    uint32_t C = (attr >> 2) & 0x1;
    uint32_t B = (attr >> 1) & 0x1;  
    trace("     attr=%x: dom=%d|xn=%d|tex=%b|C=%d|B=%d\n",
            attr, dom,xn,tex,C,B);
}

void lockdown_print_entries(const char *msg) {
    trace("-----  <%s> ----- \n", msg);
    trace("  pinned TLB lockdown entries:\n");
    for(int i = 0; i < 8; i++)
        lockdown_print_entry(i);
    trace("----- ---------------------------------- \n");
}

void pin_clear(unsigned idx) {
    return;
}

void pin_mmu_init(uint32_t domain_reg) {
    mmu_init();
    domain_access_ctrl_set(domain_reg);
}

