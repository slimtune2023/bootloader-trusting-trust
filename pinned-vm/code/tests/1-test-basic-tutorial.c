// complete working example of trivial vm:
//  1. identity map the kernel (code, heap, kernel stack, 
//     exception stack, device memory) using 1MB segments.
//  2. turn VM on/off making sure we can print.
//  3. with VM on, write to an unmapped address and make sure we 
//     get a fault.
//
//  the routines that start with "pin_" are what you'll build today,
//  the ones with with "staff_mmu_" are low level hardware 
//  routines that you'll build next VM lab.
//
//  there are some page numbers: if they start with a "B" they
//  are from the armv6 pdf, if with a number, from arm1176 pdf
#include "rpi.h"
#include "pinned-vm.h"
#include "full-except.h"
#include "memmap.h"

#define MB(x) ((x)*1024*1024)

// These are the default segments (segment = one MB)
// that need to be mapped for our binaries so far
// this quarter. 
//
// these will hold for all our tests today.
//
// if we just map these segments we will get faults
// for stray pointer read/writes outside of this region.
enum {
    // code starts at 0x8000, so map the first MB
    //
    // if you look in <libpi/memmap> you can see
    // that all the data is there as well, and we have
    // small binaries, so this will cover data as well.
    //
    // NOTE: obviously it would be better to not have 0 (null) 
    // mapped, but our code starts at 0x8000 and we are using
    // 1mb sections (which require 1MB alignment) so we don't
    // have a choice unless we do some modifications.  
    //
    // you can fix this problem as an extension: very useful!
    SEG_CODE = MB(0),

    // as with previous labs, we initialize 
    // our kernel heap to start at the first 
    // MB. it's 1MB, so fits in a segment. 
    SEG_HEAP = MB(1),

    // if you look in <staff-start.S>, our default
    // stack is at STACK_ADDR, so subtract 1MB to get
    // the stack start.
    SEG_STACK = STACK_ADDR - MB(1),

    // the interrupt stack that we've used all class.
    // (e.g., you can see it in the <full-except-asm.S>)
    // subtract 1MB to get its start
    SEG_INT_STACK = INT_STACK_ADDR - MB(1),

    // the base of the BCM device memory (for GPIO
    // UART, etc).  Three contiguous MB cover it.
    SEG_BCM_0 = 0x20000000,
    SEG_BCM_1 = SEG_BCM_0 + MB(1),
    SEG_BCM_2 = SEG_BCM_0 + MB(2),

    // we guarantee this (2MB) is an 
    // unmapped address
    SEG_ILLEGAL = MB(2),
};

// used to store the illegal address we will write.
static volatile uint32_t illegal_addr;

// macros to generate inline assembly routines.  recall
// we've used these for the debug lab, etc.
#include "asm-helpers.h"

// b4-44: get the fault address from a data abort.
cp_asm_get(cp15_fault_addr, p15, 0, c6, c0, 0)

// a trivial data abort handler that checks that we 
// faulted on the address that we expected.
//   - note: bad form that we don't actually check 
//     that its a data abort.
static void data_abort_handler(regs_t *r) {
    uint32_t fault_addr;

#if 0
    // b4-44
    // alternatively you can use the inline assembly raw.
    // can be harder to debug.
    asm volatile("MRC p15, 0, %0, c6, c0, 0" : "=r" (fault_addr));
#else
    fault_addr = cp15_fault_addr_get();
#endif

    // make sure we faulted on the address that should be accessed.
    if(fault_addr != illegal_addr)
        panic("illegal fault!  expected %x, got %x\n",
            illegal_addr, fault_addr);
    else
        trace("SUCCESS!: got a fault on address=%x\n", fault_addr);

    // done with test.
    trace("all done: going to reboot\n");
    clean_reboot();
}

void notmain(void) { 
    // ******************************************************
    // 1. one-time initialization.
    //   - create an empty page table (to catch errors).
    //   - setup exception handlers.
    //   - compute permissions for no-user-access.

    // initialize the heap.  as stated above,
    // starts at 1MB, and is 1MB big.
    kmalloc_init_set_start((void*)SEG_HEAP, MB(1));

    // example sanity check that entire process fits 
    // within 1MB segment: these symbols are in <memmap.h> 
    // and defined in <libpi/memmap> (we've used in previous
    // labs).
    assert((uint32_t)__prog_end__ < SEG_CODE + MB(1));
    assert((uint32_t)__code_start__ >= SEG_CODE);

    // setup a data abort handler (just like last lab).
    full_except_install(0);
    full_except_set_data_abort(data_abort_handler);

    // allocate a page table with invalid
    // entries.
    //
    // we will TLB pin all valid mappings.  
    // if the code is correct, the hardware
    // will never look anything up in the page
    // table.
    // 
    // however, if the code is buggy and does a 
    // a wild memory access that isn't in any
    // pinnned entry, the hardware would then try 
    // to look the address up in the page table
    // pointed to by the tlbwr0 register.
    //
    // if this page table isn't explicitly
    // initialized to invalid entries, the hardware
    // would interpret the garbage bits there 
    // valid and potentially insert them (very
    // hard bug to find).
    //
    // to prevent this we allocate a zero-filled 
    // page table.
    //  - 4GB/1MB section * 4 bytes = 4096*4.
    //  - zero-initialized will set each entry's 
    //    valid bit to 0 (invalid).
    //  - lower 14 bits (16k aligned) as required
    //    by the hardware.
    void *null_pt = kmalloc_aligned(4096*4, 1<<14);
    assert((uint32_t)null_pt % (1<<14) == 0);

    // in <mmu.h>: checks cp15 control reg 1.
    //  - will implement in next vm lab.
    assert(!mmu_is_enabled());

    // no access for user (r/w privileged only)
    // defined in <mem-attr.h>.  
    //
    // is APX and AP fields bit-wise or'd
    // together: (APX << 2 | AP) see 
    //  - 3-151 for table, or B4-9
    enum { no_user = perm_rw_priv };

    // armv6 has 16 different domains with their own privileges.
    // just pick one for the kernel.
    enum { dom_kern = 1 };

    // attribute for device memory (see <pin.h>).  this 
    // is needed when pinning device memory:
    //   - permissions: kernel domain, no user access, 
    //   - memory rules: strongly ordered, not shared.
    //     see <mem-attr.h> for <MEM_device>
    pin_t dev  = pin_mk_global(dom_kern, no_user, MEM_device);

    // attribute for kernel memory (see <pin.h>)
    //   - protection: same as device; we don't want the
    //     user to read/write it.
    //   - memory rules: uncached access.  you can start
    //     messing with this for speed, though have to 
    //     do cache coherency maintance
    pin_t kern = pin_mk_global(dom_kern, no_user, MEM_uncached);

    // Q1: if you uncomment this, what happens / why?
    // kern = pin_mk_global(dom_kern, perm_ro_priv, MEM_uncached);

    // ******************************************************
    // 2. map the process's memory.  We use the enums
    //    above that define where everything is.

    // <mmu.h>: one-time initialize of MMU hardware
    //  - invalidate TLB, caches, etc.
    //  - will implement in next vm lab.
    mmu_init();

    printk("Before first iter:\n");
    // lockdown_print_entries("");

    // identiy map all segments in one of the available
    // 0..7 TLB pinned entries.
    pin_mmu_sec(0, SEG_CODE, SEG_CODE, kern);

    // printk("After iter 1:\n");
    // lockdown_print_entries("");

    pin_mmu_sec(1, SEG_HEAP, SEG_HEAP, kern); 

    // Q2: if you comment this out, what happens?
    pin_mmu_sec(2, SEG_STACK, SEG_STACK, kern); 
    // Q3: if you comment this out, what happens?
    pin_mmu_sec(3, SEG_INT_STACK, SEG_INT_STACK, kern); 

    // map all device memory to itself.  ("identity map")
    pin_mmu_sec(4, SEG_BCM_0, SEG_BCM_0, dev); 
    pin_mmu_sec(5, SEG_BCM_1, SEG_BCM_1, dev); 
    pin_mmu_sec(6, SEG_BCM_2, SEG_BCM_2, dev); 

    printk("At end:\n");
    // lockdown_print_entries("");

    // ******************************************************
    // 3. setup virtual address context.
    //  - domain permissions.
    //  - page table, asid, pid.

    // b4-42: give permissions for all domains.
    // Q4: if you set this to ~0, what happens w.r.t. Q1?
    // Q5: if you set this to 0, what happens?
    domain_access_ctrl_set(DOM_client << dom_kern*2); 
    // staff_domain_access_ctrl_set(~0); 

    // set address space id, page table, and pid.
    // note:
    //  - <pid> is ignored by the hw: it's just to help the os.
    //  - <asid> doesn't matter for this test b/c all entries 
    //    are global.
    //  - recall the page table has all entries invalid and is
    //    just to catch memory errors.
    enum { ASID = 1, PID = 128 };
    mmu_set_ctx(PID, ASID, null_pt);

    // if you want to see the lockdown entries.
    // lockdown_print_entries("about to turn on first time");

    // ******************************************************
    // 4. turn MMU on/off, checking that it worked.
    trace("about to enable\n");
    for(int i = 0; i < 10; i++) {
        assert(!mmu_is_enabled());
        // mmu.h: you'll implement next vm lab.  
        mmu_enable();

        // this uses: stack, code, data, BCM.
        if(mmu_is_enabled())
            trace("MMU ON: hello from virtual memory!  cnt=%d\n", i);
        else
            panic("MMU is not on?\n");

        // mmu.h: you'll implement next vm lab.  
        mmu_disable();
        assert(!mmu_is_enabled());
        trace("MMU is off!\n");
    }

    // ******************************************************
    // 5. setup exception handling and make sure we get a fault.
    // 
    //    you should make a copy of this test case, and 

    // we write to an address we know is unmapped so we get
    // a fault.
    illegal_addr = SEG_ILLEGAL;

    // this <PUT32> should "work" since vm is off.
    assert(!mmu_is_enabled());
    PUT32(illegal_addr, 0xdeadbeef);
    trace("we wrote without vm: got %x\n", GET32(illegal_addr));
    assert(GET32(illegal_addr) == 0xdeadbeef);

    // now try with MMU on: it's a good exercise to make a copy
    // of this test case and change it so you repeatedly fault.
    mmu_enable();
    assert(mmu_is_enabled());
    output("about to write to <%x>\n", illegal_addr);
    // this <PUT32> should cause a data_abort.
    PUT32(illegal_addr, 0xdeadbeef);
    panic("should not reach here\n");
}
