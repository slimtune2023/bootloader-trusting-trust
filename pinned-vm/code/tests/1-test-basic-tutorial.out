TRACE:notmain:about to enable
TRACE:notmain:MMU ON: hello from virtual memory!  cnt=0
TRACE:notmain:MMU is off!
TRACE:notmain:MMU ON: hello from virtual memory!  cnt=1
TRACE:notmain:MMU is off!
TRACE:notmain:MMU ON: hello from virtual memory!  cnt=2
TRACE:notmain:MMU is off!
TRACE:notmain:MMU ON: hello from virtual memory!  cnt=3
TRACE:notmain:MMU is off!
TRACE:notmain:MMU ON: hello from virtual memory!  cnt=4
TRACE:notmain:MMU is off!
TRACE:notmain:MMU ON: hello from virtual memory!  cnt=5
TRACE:notmain:MMU is off!
TRACE:notmain:MMU ON: hello from virtual memory!  cnt=6
TRACE:notmain:MMU is off!
TRACE:notmain:MMU ON: hello from virtual memory!  cnt=7
TRACE:notmain:MMU is off!
TRACE:notmain:MMU ON: hello from virtual memory!  cnt=8
TRACE:notmain:MMU is off!
TRACE:notmain:MMU ON: hello from virtual memory!  cnt=9
TRACE:notmain:MMU is off!
TRACE:notmain:we wrote without vm: got 0xdeadbeef
TRACE:data_abort_handler:SUCCESS!: got a fault on address=0x200000
TRACE:data_abort_handler:all done: going to reboot
